pylzy/tests/env/test_provisioning.py

```diff
@@ -0,0 +1,166 @@
+import pytest
+
+import dataclasses
+
+from lzy.types import VmSpec
+from lzy.env.base import NotSpecified
+from lzy.env.provisioning.provisioning import Provisioning, Any, NO_GPU
+from lzy.env.provisioning.score import maximum_score_function, minimum_score_function
+from lzy.exceptions import BadProvisioning
+
+
+@pytest.fixture
+def first():
+    return Provisioning(cpu_count=1, gpu_count=Any)
+
+
+@pytest.fixture
+def second():
+    return Provisioning(cpu_type='foo', gpu_type=Any, gpu_count=2)
+
+
+def test_combine(first, second):
+    assert first.combine(second) == Provisioning(
+        cpu_count=1,
+        cpu_type='foo',
+        gpu_count=2,
+        gpu_type=Any
+    )
+
+    assert second.combine(first) == Provisioning(
+        cpu_count=1,
+        cpu_type='foo',
+        gpu_count=Any,
+        gpu_type=Any,
+    )
+
+
+def test_validate(first, second):
+    with pytest.raises(BadProvisioning):
```

Comment:
you specify some meaningful exception message, i think it's better to check it in tests like so:
```
with pytest.raises(BadProvisioning) as e:
    Provisioning(gpu_count=1).validate()
assert str(e.value) == 'your message`
```
by this we can check that user will see